common/src/jni/main/cpp/conscrypt/native_crypto.cc

```diff
@@ -7048,6 +7048,97 @@ static jbyteArray NativeCrypto_SSL_get_tls_unique(JNIEnv* env, jclass, jlong ssl
     return byteArray.release();
 }
 
+static void NativeCrypto_SSL_set_token_binding_params(JNIEnv* env, jclass, jlong ssl_address,
+        CONSCRYPT_UNUSED jobject ssl_holder, jintArray params) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params", ssl);
+    if (ssl == nullptr) {
+        return;
+    }
+    ScopedIntArrayRO paramsValues(env, params);
+    int ret;
+    if (paramsValues.get() == nullptr) {
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params params==null", ssl);
+        ret = SSL_set_token_binding_params(ssl, nullptr, 0);
+    } else {
+        std::unique_ptr<uint8_t[]> paramsBytes(new uint8_t[paramsValues.size()]);
+        for (int i = 0; i < paramsValues.size(); i++) {
+            paramsBytes.get()[i] = static_cast<uint8_t>(paramsValues[i]);
+        }
+        ret = SSL_set_token_binding_params(ssl, paramsBytes.get(), paramsValues.size());
+    }
+
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params => %d", ssl, ret);
+
+    if (!ret) {
+        conscrypt::jniutil::throwSSLExceptionStr(env, "Could not set token binding parameters");
+        ERR_clear_error();
+    }
+}
+
+static int NativeCrypto_SSL_get_token_binding_params(JNIEnv* env, jclass, jlong ssl_address,
+        CONSCRYPT_UNUSED jobject ssl_holder) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_get_token_binding_params", ssl);
+    if (ssl == nullptr) {
+        return 0;
+    }
+    int ret;
+    if (!SSL_is_token_binding_negotiated(ssl)) {
+        ret = -1;
+    } else {
+        ret = SSL_get_negotiated_token_binding_param(ssl);
+    }
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_set_token_binding_params => %d", ssl, ret);
+    return ret;
+}
+
+static jbyteArray NativeCrypto_SSL_export_keying_material(JNIEnv* env, jclass, jlong ssl_address,
+        CONSCRYPT_UNUSED jobject ssl_holder, jbyteArray label, jbyteArray context, jint num_bytes) {
+    CHECK_ERROR_QUEUE_ON_RETURN;
+    SSL* ssl = to_SSL(env, ssl_address, true);
+    JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material", ssl);
+    if (ssl == nullptr) {
+        return nullptr;
+    }
+    ScopedByteArrayRO labelBytes(env, label);
+    if (labelBytes.get() == nullptr) {
+        JNI_TRACE("ssl=%p NativeCrypto_SSL_export_keying_material label == null => exception", ssl);
+        return nullptr;
+    }
+    std::unique_ptr<uint8_t[]> out(new uint8_t[num_bytes]);
+    int ret;
+    if (context == nullptr) {
+        ret = SSL_export_keying_material(ssl, out.get(), num_bytes,
+                        reinterpret_cast<const char*>(labelBytes.get()), labelBytes.size(),
+                        nullptr, 0, 0);
+    } else {
+        ScopedByteArrayRO contextBytes(env, context);
+        if (contextBytes.get() == nullptr) {
```

Comment:
remind me ... can this happen if `context != nullptr`?

Reply:
I doubt it, but the JNI docs for GetFooArrayElements() only say "Returns a pointer to the array elements, or NULL if the operation fails.", so better safe than sorry.

Reply:
(The docs also say "Since the returned array may be a copy of the Java array, changes made to the returned array will not necessarily be reflected in the original array until Release<PrimitiveType>ArrayElements() is called.", so the failure case was probably intended to be if allocating the copy failed.)